### ISSUE TITLE
keyscript not being run for secondary devices

### ISSUE THREAD

**Issue by kelderek** [kelderek/TPM2-LUKS#2]
I tried adding some drives to a VM, and manually encrypting them.  I have added them to /etc/fstab and /etc/crypttab, and my script seems them and seems to work properly, but on reboot it doesn't appear the keyscript is being run, I just see the systemd prompt for a password (not the one from tpm2-getkey).  Because systemd doesn't support the keyscript parameter, my interpretation is that initrd is not bothering to mount those drives before starting systemd.  Maybe because I am mounting them at /mnt so they aren't important for boot?

@ezaton Any thoughts on this or how I can make sure they get mounted by initrd?  It is entirely possible I did something wrong in fstab or crypttab, but I tried to mirror the encrypted root partition as closely as possible...

**Comment by ezaton** [comment]
Did you try a clean Ubuntu install, specifying both disks are members of the same VG? (both as encrypted LUKs devices)

**Comment by kelderek** [comment]
It is a clean install, but the disks are not part of the same VG.  They aren't in LVM at all, just disks added, then manually LUKS encrypted.  Adding them to the same VG would probably fix it, I will see about doing that.  I wonder if adding them to their own VG would be enough, I will give that a try too.  Ideally this would work regardless of VG, but that may not be feasible.  It would be a reasonable alternative in that case to to just store a keyfile on the main drive to unlock the others.  Thanks!

**Comment by ezaton** [comment]
I think you are targeting "all possible scenarios" which is not what you should do. The initramfs (initrd) should handle the minimal disks required to boot, which means - /. Other encrypted disks would be decrypted and mounted later, and this is a different flow you need to hack to use TPM. If all encrypted disks are members of the same VG, you are forced to decrypt them on-boot, before you can activate the VG, and thus - the logic of hooking into initramfs. Otherwise - you should investigate systemd service and hack its passphrase query call into something you can handle through TPM.
So - to sum things up: Either all of your disks are members of the same VG containing / and should be decrypted on pre-root phase, or some of them are not, and would be relevant later (maybe) via systemd

**Comment by plant7403** [comment]
Omg, I thought only I was unlucky and it wasn't working for me
I have two disks, one the system, and the other one is data

**Comment by kelderek** [comment]
Haha yeah it isn't just you.  initrd boots just enough to pass off to systemd, and systemd doesn't support keyscripts for some reason.  I saw some stuff here: https://github.com/systemd/systemd/pull/3007 that might let this work with systemd but I haven't had the time to look further into it

**Comment by kelderek** [comment]
@xxxcrow I tinkered around a bit last night and working with file sockets and the systemd socket services is new to me so I don't have it working with the TPM yet.  In the meantime you could use keyfiles stored on your system partition to automatically unlock your data drives, which should be safe enough if your system partition is also encrypted.

You can customize the locations if needed, but I used the following method:
1) Create the keyfile(s) at **/etc/cryptsetup-keys.d/_volume_.key** e.g. for /dev/sdb1 the file would be /etc/cryptsetup-keys.d/sdb1.key and the contents is just the key and _nothing else_, not even a newline.  e.g. created with echo -n 'yourencryptionpassword' > /etc/cryptsetup-keys.d/sdb1.key
2) Set the permissions - systemd will refuse to use it if permissions are too lax.  chmod 400 /etc/cryptsetup-keys.d/sdb1.key

That should be it.  As long as the /etc/crypttab entry for that partition has the third column empty or set to none or -, systemd will look for the file at /etc/cryptsetup-keys.d/_volume_.key and /run/cryptsetup-keys.d/_volume_.key.  You should also be able to change the third column to the full path of the key if you want to store it in an alternative location.

**Comment by plant7403** [comment]
@kelderek I will try it tomorrow, this should work
I thought about this way, but for some reason mounting on boot seemed more logical/secure
And I still don't understand why keyscript wouldn't have worked 🥲

**Comment by kelderek** [comment]
Using the keyfiles this way can still mount on boot.  Until I get it updated to use the TPM to properly seal the key based on a combination of PCR values, the keyfiles on an encrypted root should be just as good for your data volumes.  Better if you don't have your BIOS password protected with USB boot options disabled.

As for why the keyscript doesn't work it is because the systemd devs didn't like keyscripts, so when systemd takes over mounting drives after initrd mounts the root partition, it ignores the keyscript.  They provided an alternate path forward by making a systemd socket and service that so an AF_UNIX file socket can be called which _then_ runs the keyscript.  I am tinkering with that to see if I can get it working. A whole lot of obfuscation just to avoid running the keyscript directly...

**Comment by plant7403** [comment]
@kelderek

> Better if you don't have your BIOS password protected with USB boot options disabled.

??? 😬

**Comment by kelderek** [comment]
Yikes, bad typo on my part.  I meant the exact opposite, you SHOULD have your bios password protected and USB boot disabled if you are using my script.

**Comment by plant7403** [comment]
@kelderek by the way, for some reason it's not working for me 🫡

**Comment by kelderek** [comment]
@xxxcrow I found that adding initramfs as part of the /etc/crypttab parameters forces initrd to unlock the volume instead of passing it off to systemd.  If you would like to give it a try, check out the alpha version:
https://github.com/kelderek/TPM2-LUKS/tree/v0.8.0-alpha.1

It doesn't have the command line parameters but it does have a menu system to pick the drive(s) you want.  It is working for me with 2 encrypted drives.  I would love to hear your feedback!
